app/views/teachers/classes/CourseProgressDotView.vue

```diff
@@ -0,0 +1,266 @@
+                    {{ percentCompleted }}%
+                    {{ $t('teacher.completed') }}
+                </div>
+            </template>
+        </v-popover>
+    </li>
+</template>
+
+<script>
+  import { mapState } from 'vuex'
+
+  export default {
+    props: {
+      course: Object,
+      classroom: Object,
+    },
+
+    computed: Object.assign({},
+      // TODO this could be loading (top level component prevents this now but may not in future).  Handle loading state here
+      mapState('courses', {
+        courses: function (state) {
+          return state.byId
+        }
+      }),
+
+      mapState('courseInstances', {
+        // TODO this could be loading (top level component prevents this now but may not in future).  Handle loading state here
+        courseInstancesLoading (state) {
+          return state.loading.byTeacher[this.$props.classroom.ownerID]
+        },
+
+        courseInstance (state) {
+          const instances = state.courseInstancesByTeacher[this.$props.classroom.ownerID] || []
+
+          return instances
+            .find(i => i.courseID === this.$props.course._id && i.classroomID === this.$props.classroom._id)
+        }
+      }),
+
+      mapState('levelSessions', {
+        // TODO this could be loading (top level component prevents this now but may not in future).  Handle loading state here
+        levelSessionsLoading (state) {
+          return state.loading.sessionsByClassroom[this.$props.classroom._id]
+        },
+
+        levelCompletionsByUser (state) {
+          const levelSessionState = state.levelSessionsByClassroom[this.$props.classroom._id]
+
+          if (levelSessionState) {
+            return levelSessionState.levelCompletionsByUser
+          }
+
+          return {}
+        }
+      }),
+
+      {
+        // TODO course acronym could be controlled and sent from the backend
+        courseAcronym: function () {
+          const course = this.courses[this.$props.course._id]
+
+          let prefix = 'CS';
+          if (/game-dev/.test(course.slug)) {
+            prefix = 'GD'
+          } else if (/web-dev/.test(course.slug)) {
+            prefix = 'WD'
+          }
+
+          let number = '1';
+          const numberMatch = (course.slug || '').match(/(\d+)$/)
+          if (numberMatch && numberMatch[1]) {
+            number = numberMatch[1]
+          }
+
+          return `${prefix}${number}`
+        },
+
+        // TODO this should be moved to a stats store, calculated on the backend, or some combination of the two
+        courseLevels: function () {
+          const classroom = this.$props.classroom
+          const course = (classroom.courses || []).find(c => c._id === this.$props.course._id)
+
+          if (!course || !course.levels) {
+            return []
+          }
+
+          return course.levels
+              .filter(l => !l.practice && !l.assessment)
+              .map(l => l.original)
+        },
+
+        courseStats: function () {
```

Comment:
We have some implementation for `@getLevels` and `statsForSessions` here:
https://github.com/codecombat/codecombat/blob/master/app/models/Classroom.coffee#L78
https://github.com/codecombat/codecombat/blob/master/app/models/Classroom.coffee#L121

Looks like we are doing something similar here but not sure if backbone models are available here so that we can re-use that implementation. Otherwise, can we move this to a re-usable place (maybe modules) for future Vue pages?

Reply:
Jotting some notes here / going to think more on this tonight:

Definitely.  For now I think I'll add comments to make sure the relationship is clear.  The plan is to move this code into the stats jobs that we discussed today so that it can be completely shared.  In fact this code can be dropped right into the repo for calculation.  The blocker is understanding the level session loading / performance implications.  Until I understand that better I don't feel comfortable doing this and I don't think we want to invest the time right now.  If anyone feels comfortable doing so it may be a great pair programming opportunity to quickly remove this code and transition to a shared background calculation.

Reply:
Added clarifying comments